extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -0,0 +1,200 @@
+        }
+      }
+    }
+    const replacements = urlReplacementsForDoc(this.getAmpDoc());
+    return url => replacements.expandUrlSync(
+        url, vars, undefined /* opt_collectVars */, whitelist);
+  }
+
+  /**
+   * Attemts to issue a request to `url` to report the click. The request method
+   * depends on the exit config's transport property. navigator.sendBeacon will
+   * be tried if transport.beacon is `true` or `undefined`. Otherwise, or if
+   * sendBeacon returns false, an image request will be made.
+   * @param {string} url
+   */
+  pingTrackingUrl_(url) {
+    user().fine(TAG, `pinging ${url}`);
+    const useBeacon = this.config_.transport[TransportMode.BEACON] == true ||
+        !this.config_.transport.hasOwnProperty(TransportMode.BEACON);
+    if (useBeacon &&
+        this.win.navigator.sendBeacon &&
+        this.win.navigator.sendBeacon(url, '')) {
+      return;
+    }
+    const req = this.win.document.createElement('img');
+    req.src = url;
+  }
+
+  /**
+   * Checks the click event against the given filters. Returns true if the event
+   * passes.
+   * @param {!Array<string>} names
+   * @param {!Object<string, !./config.FilterConfig>} specs
+   * @param {!Event} event
+   * @returns {boolean}
+   */
+  filter_(names, specs, event) {
+    if (!names) { return true; }
+    return names.every(name => {
+      const spec = specs[name];
+      if (!spec) {
+        user().warn(TAG, `Filter '${name}' not found`);
+        return true;
+      }
+      const filter = this.findFilterForType_(spec.type);
+      if (!filter) {
+        user().warn(
+            TAG, `Invalid filter type '${spec.type}' for filter '${name}'`);
+        return true;
+      }
+      const result = filter.filter(spec, event);
+      user().info(TAG, `Filter '${name}': ${result ? 'pass' : 'fail'}`);
+      return result;
+    });
+  }
+
+  /** @return {?./filters/filter.Filter} */
+  findFilterForType_(type) {
+    switch (type) {
+      case FilterType.CLICK_DELAY:
+        return this.clickDelayFilter_;
+      case FilterType.CLICK_LOCATION:
+        // TODO(clawr): implement this.
+      default:
+        return null;
+    }
+  }
+
+  /** @override */
+  buildCallback() {
+    this.element.setAttribute('aria-hidden', 'true');
+
+    try {
+      const children = this.element.children;
+      if (children.length != 1) {
+        throw new Error('The tag should contain exactly one <script> child.');
+      }
+      const child = children[0];
+      if (isJsonScriptTag(child)) {
+        this.config_ = assertConfig(JSON.parse(child.textContent));
+      } else {
+        throw new Error(
+            'The amp-ad-exit config should ' +
+            'be put in a <script> tag with type="application/json"');
+      }
+    }
+    catch (e) {
+      user().error(TAG, 'Invalid JSON config', e);
+      throw e;
+    }
+  }
+
+  /** @override */
+  viewportCallback(inViewport) {
+    if (inViewport) {
+      this.clickDelayFilter_.resetClock();
```

Comment:
What kind of layout does this element have? Non-displayed elements won't ever have `#viewportCallback` called.